src/server/pfs/server/driver.go

```diff
@@ -4539,14 +4545,20 @@ func (s *putFileServer) Peek() (*pfs.PutFileRequest, error) {
 	return req, nil
 }
 
-func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) (oneOff bool, repo string, branch string, err error) {
+func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) (bool, string, string, error) {
+	// return values
+	var oneOff bool
+	var repo string
+	var branch string
+
 	var pr *io.PipeReader
 	var pw *io.PipeWriter
```

Comment:
The code here is kinda convoluted, but I think each `io.Pipe()` call has a matching `pw.Close()` call. There's two close sites, one that happens before we assign a new value to `pw` and one that happens at the end no matter what. That one is actually a little confusing, because it actually calls `pw.CloseWithError(err)`, which makes it look like it's handling an error, rather than normal execution, but under normal execution `err` will be equal to `io.EOF` and `pw.Close()` is actually just implemented as `pw.CloseWithError(io.EOF)` so I simplified it a bit.